Add sensors to honeywell

cfairfax opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This change adds a sensor platform to the honeywell integration component. With this change, all indoor and outdoor temperature and humidity sensors will be available as entities.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

climate:
   - platform: honeywell
     username: tcc_username
     password: tcc_password


```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->
In order to make the climate and sensor platform share the same data, I needed to make some structural changes to the climate platform. It will now use DataUpdateCoordinator to fetch data instead of fetching data for each device. I also needed to move the login and update code out of climate.py and into __init__.py


- This PR fixes or closes issue: fixes #
https://community.home-assistant.io/t/honeywell-thermostat-outside-temp-and-humidity/150267
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/15414

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass** (All the honeywell tests are being skipped)
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

cfairfax committed:
initial submit of honeywell changes
convert stuff to async
use single api poll to make fewer api requests
add sensors for indoor/outdoor temperature and humidity

cfairfax committed:
initial submit of honeywell changes
convert stuff to async
use single api poll to make fewer api requests
add sensors for indoor/outdoor temperature and humidity

cfairfax committed:
Merge branch 'dev' of https://github.com/cfairfax/core into dev

cfairfax committed:
first check to make sure the sensors are available before adding

cfairfax committed:
wrap somecomfort api calls in a neat async_wrapper to make it a good asyncio citizen

cfairfax committed:
add better exception handling so that the component can recover from api errors

cfairfax committed:
remove unnecessary import

cfairfax committed:
Merge remote-tracking branch 'upstream/dev' into dev

cfairfax committed:
format code

cfairfax committed:
Update __init__.py

change update interval to 5 minutes

cfairfax committed:
Merge branch 'dev' of https://github.com/cfairfax/core into dev

cfairfax committed:
fix lint issues and formatting, add docstrings

cfairfax committed:
Merge branch 'dev' into add_sensors_to_honeywell

cfairfax committed:
Merge remote-tracking branch 'upstream/dev' into dev

homeassistant commented:
Hi @cfairfax, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#42381).

Once you do that we will be able to review and accept this pull request.

Thanks!

cfairfax committed:
add new files to coveragerc

cfairfax committed:
Merge branch 'add_sensors_to_honeywell' of https://github.com/cfairfax/core into add_sensors_to_honeywell

homeassistant commented:
Hi @cfairfax, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#42381).

Once you do that we will be able to review and accept this pull request.

Thanks!

cfairfax committed:
fix constant naming style

make CLIENT_KEY_COORDINATOR all caps

cfairfax committed:
Merge branch 'add_sensors_to_honeywell' of https://github.com/cfairfax/core into add_sensors_to_honeywell

cfairfax committed:
fix pylint issues
instead of getting at the json returned by the honeywell api, use
the somecomfort functions

cfairfax commented:
Hi @zxdavb, let me know what you think. This will be my first contribution to an open source project, so I'm expecting there will be some feedback before approval. 

Do I need to add anything to the PR description? 

Also, should I collapse the commits into one? There ended up being more than I expected.

frenck commented on homeassistant/components/honeywell/sensor.py:
```diff
@@ -0,0 +1,150 @@
+
+PLATFORM_SCHEMA = vol.All(
+    cv.deprecated(CONF_REGION),
+    PLATFORM_SCHEMA.extend(
```
We no longer accept YAML configuration for integrations that connect to devices or services.

Instead, a configuration flow via the UI should be used.

For more information about this decision see:

- ADR-0007: <https://github.com/home-assistant/architecture/blob/master/adr/0007-integration-config-yaml-structure.md#decision>
- ADR-0010: <https://github.com/home-assistant/architecture/blob/master/adr/0010-integration-configuration.md#decision>

To get started on the integration configuration, see our developer documentation:

<https://developers.home-assistant.io/docs/config_entries_config_flow_handler>

cfairfax replied:
Thanks for getting back to me. I think for this change I will try using discovery.load_platform instead of modifying the yaml structure.
`hass.helpers.discovery.load_platform("sensor", "honeywell", {}, config)`
Would that work here?

I would be open to converting this to use config entries in the future, but at this point I think that's too big of a change for me.

MartinHjelmare replied:
We don't allow new platform schemas. This needs a config flow and a major refactor. So unfortunately it'll be quite a big task.

If you don't have time for that, let's close this PR.

If you do want to work on config flow, one way of allowing more time for it, is to break up the PR into smaller pieces and use a separate target branch instead of dev branch, to allow multiple PRs before releasing everything together.

Please let us know your intentions.

cfairfax committed:
use discovery.async_load_platform to init the sensors
this means existing honeywell users will not need to update their configuration.yaml
also fix somecomfort usage in sensor.py

cfairfax committed:
remove throwaway print

cfairfax commented:
I found a stability issue I want to fix. Let me get back to you after I try it on my system for a week

cfairfax committed:
add back code for automatic retry if update failed

cfairfax committed:
don't catch generic Exception type

cfairfax commented:
The latest commit fixes the stability issue. I assume this is way too many commits to be merged directly. Should I combine them into a single commit?

cfairfax commented:
I notice that the integration still seems to get into a bad state from time to time. I noticed it's whenever the DataUpdateCoordinator's update_method raises UpdateFailed exception. I can see in the log that it "recovers", but the sensors stop updating in lovelace history panel after the first UpdateFailed exception is raised.

I see this line printed:
2020-11-21 11:58:25 INFO (MainThread) [custom_components.honeywell] Fetching honeywell_update_coordinator data recovered
But the sensors stop updating. Is there anything else that needs to happen in order for it to recover?

cfairfax committed:
fix bug where the coordinator would go stale when data fetching failed

cfairfax committed:
Merge branch 'dev' of https://github.com/home-assistant/core into add_sensors_to_honeywell

cfairfax committed:
add another exception to catch
